apps/src/javalab/JavabuilderConnection.js

```diff
@@ -84,12 +122,23 @@ export default class JavabuilderConnection {
       // event.code is usually 1006 in this case
       console.log(`[close] Connection died. code=${event.code}`);
     }
+  }
+
+  onExit() {
     if (this.miniApp) {
       // miniApp on close should handle setting isRunning state as it
       // may not align with actual program execution. If mini app does
       // not have on close we won't toggle back automatically.
+      // We also pass onOutputMessage so the mini app can send its own custom
+      // done message.
```

Comment:
Where is this being passed?

Reply:
whoops, I originally was passing this but decided to switch to using redux directly instead. Per your other comment, I think it makes sense to switch to using redux directly for everything, now that we need more than one redux method.

Reply:
thinking more on this, I think I will pass in the redux abstractions instead to make testing easier

Reply:
pushed update that refactors to pass in all redux functions